### PR TITLE
Add quality gate failure messages to the report

### PR DESCRIPTION
**Pull request by bgavrilMS**
I like pickles.

### CODE REVIEW

File: Tasks/Common/SonarQubeHelpers/SonarQubeHelper.ps1 [hunk 1]
```diff
@@ -17,14 +17,17 @@ function SetTaskContextVariable
 {
     param([string][ValidateNotNullOrEmpty()]$varName, 
           [string]$varValue)
-    
+        
+    #[Environment]::SetEnvironmentVariable($varName, $varValue)
```
Comment (bgavrilMS) [inline Tasks/Common/SonarQubeHelpers/SonarQubeHelper.ps1]: I'll remove these.

File: Tasks/SonarQubePostTest/PRCA/Orchestrator.ps1 [hunk 2]
```diff
@@ -19,6 +19,10 @@ function HandleCodeAnalysisReporting
         
         PostAndResolveComments $messages "SonarQube Code Analysis"
     }	
+    else
+    {
+        Write-Verbose "The build was not triggered by a Pull Request, not processing code analysis comments"
```
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/PRCA/Orchestrator.ps1]: Unrelated change - there should be a log message in the build is not a PR build to diagnose potential issues.

File: Tasks/SonarQubePostTest/SonarQubePostTest.ps1 [hunk 3]
```diff
@@ -12,13 +12,11 @@ ExitOnPRBuild
 . $PSScriptRoot/PRCA/Orchestrator.ps1
 
 . $PSScriptRoot/SonarQubeMetrics.ps1
-. $PSScriptRoot/SonarQubeReportHandler.ps1
+. $PSScriptRoot/SummaryReport/ReportBuilder.ps1
```
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SonarQubePostTest.ps1]: I split the summary report logic into multiple files and placed them into a SummaryReport directory

File: Tasks/SonarQubePostTest/SummaryReport/DashboardUriSection.ps1 [hunk 4]
```diff
@@ -0,0 +1,22 @@
+function GetDashboardUriSectionContent
```
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SummaryReport/DashboardUriSection.ps1]: If I were using a normal language like C# I would have used some aggregator sort of pattern. But this is PS , so it will have to be done by convention - the top level report builder calls each Get_Section function to construct the report and then combines the sections.

File: Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1 [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    $section1 = FormatQualityGateStatusSection $qualityGateStatus
+    
+    $warningsAndErrors = GetQualityGateWarningsAndErrors $qualityGateResponse    
+    $section2 = FormatQualityGateMessagesSection $warningsAndErrors
+    
+    $content = FormatQualityGateSection $section1 $section2
+    return $content
+}
+
+function FormatQualityGateSection
```
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1]: I will rename section1 and section2 to qualityGateStatusSection and qualityGateMessagesSection

File: Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1 [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    
+    $reportContents = [String]::Format($template, $qualityGateColor, $qualityGateLabel)
+    return $reportContents
+}
+
+#
+# Fetches the quality gate status and the errors and warnings and also the metrics in order to provide a user friendly report.
+# Note: this is method is unit tested
+#
+function GetQualityGateWarningsAndErrors
```
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1]: This is the where the "view model" gets constructed based on data from the SQ server

File: Tasks/SonarQubePostTest/SummaryReport/ReportBuilder.ps1 [hunk 7]
```diff
@@ -59,33 +61,26 @@ function IsReportEnabled
...
 {
-    param($reportPath)
+    $content = BuildReportContent
+    $reportPath = CreateReportFile $content
     
-    Write-Host "##vso[task.addattachment type=Distributedtask.Core.Summary;name=SonarQube Analysis Report;]$reportPath"
+    return $reportPath
 }
 
-function CreateReport
```
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SummaryReport/ReportBuilder.ps1]: All the HTML specific bits were moved to the section builders

File: Tests/L0/SonarQubePostTest/SummaryReport.ps1 [hunk 8]
```diff
@@ -53,16 +64,31 @@ Register-Mock IsPrBuild {$false}
...
 $file.Dispose()
 [IO.Directory]::Delete($tempDir, $true)    
-Unregister-Mock  GetSonarQubeOutDirectory
-Unregister-Mock  Write-Host 
+Unregister-Mock GetSonarQubeOutDirectory
+Unregister-Mock Write-Host 
 Unregister-Mock IsPrBuild   
 
 
+### Test 3 - Fetching the quality gate warnings and errors
```
Comment (bgavrilMS) [inline Tests/L0/SonarQubePostTest/SummaryReport.ps1]: I don't see the value in having tests that compare HTML so I added a unit test for the logic that forms the "view model"

File: Tests/L0/SonarQubePostTest/data/ReportTest/metricsResponse.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            "id": "108",
+            "key": "sqale_index",
+            "type": "WORK_DUR",
+            "name": "Technical Debt"
+        },
+        {
+            "id": "109",
+            "key": "new_technical_debt",
+            "type": "WORK_DUR",
```
Comment (bgavrilMS) [inline Tests/L0/SonarQubePostTest/data/ReportTest/metricsResponse.json]: Note that I could add the unit of measurement to the metrics. However I fear we will be too closely coupled with SonarQube in this case. I will discuss with Jean-Marc further.

File: Tasks/SonarQubePostTest/SonarQubeMetrics.ps1 [hunk 10]
```diff
@@ -38,18 +38,38 @@ function WaitForAnalysisToFinish
     Assert (![String]::IsNullOrEmpty($analysisId)) "Could not fetch the analysis id"
 }
 
+
+function FetchQualityGateDetails
+{
+    param ([Array]$errors,[Array]$warnings)
+    
+    $analysisId = GetTaskContextVariable "MSBuild.SonarQube.AnalysisId"             
+    Assert (![String]::IsNullOrEmpty($analysisId)) "WaitForAnalysisToFinish should be called first."
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: If you're fetching id in multiple places then it might be worth putting lines 46 and 47 into a separate function.

File: Tasks/SonarQubePostTest/SonarQubeMetrics.ps1 [hunk 11]
```diff
@@ -38,18 +38,38 @@ function WaitForAnalysisToFinish
...
+       
+    $response = InvokeGetRestMethod "/api/qualitygates/project_status?analysisId=$analysisId" $true
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: How are different versions of the SQ server being handled (specifically older versions that don't have this API)?
Is there a minimum level of SQ server that is required for the quality gate summary to appear?

File: Tasks/SonarQubePostTest/SonarQubeMetrics.ps1 [hunk 12]
```diff
@@ -38,18 +38,38 @@ function WaitForAnalysisToFinish
...
+    return $response
+}
+
+#
+# For all the metrics in the system, this method returns the key and the friendly name, as well as the type and the id   
+#
+function FetchMetricNames
+{
+    $response = InvokeGetRestMethod "/api/metrics/search?ps=500&f=name" $false
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: The docs say ps must be less then 500

File: Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1 [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+            [void]$sb.AppendLine("<td><span style=""padding-right:4px;"">$($message.metric_name)</span></td>")
+            [void]$sb.AppendLine("<td style=""text-align: center; background-color:$($message.color); color:#fff;""><span style=""padding:0px 2px"">$($message.actualValue)</span></td>")
+            [void]$sb.AppendLine('<td>&nbsp;' + $message.comparator + ' ' + $message.threshold + '</td>')
+            [void]$sb.AppendLine("</tr>")
+        }
+        
+        [void]$sb.AppendLine("</table>")
+    } 
+    
+    return $sb.ToString()
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1]: It looks like $sb won't be initialised unless there are message (line 33). Won't this call fail in that case?

File: Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1 [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+function FormatQualityGateSection
+{
+    param ($section1, $section2)
+    
+    return "$section1 $section2"
+}
+
+function FormatQualityGateMessagesSection
+{
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1]: It's not easy to see what is supposed to be in each section: a short comment giving an example of the what the section is supposed to contain would help e.g.

Example output:
Duplicated blocks 1 < 10
Technical debt on new code  99 < 200

File: Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1 [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+            {
+                return '&#60;'
+            }
+        {$_ -eq "NE"} 
+            {
+                return '&#8800;'
+            }
+        Default 
+        {
+            Write-Warning "Unkown operator $comparator"
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SummaryReport/QualityGateSection.ps1]: Typo

File: Tasks/SonarQubePostTest/SummaryReport/ReportBuilder.ps1 [hunk 16]
```diff
@@ -95,45 +90,12 @@ function CreateReport
...
-        {
-            Write-Warning "Could not detect the quality gate status or a new status has been introduced."
-            $color.Value = "#bbb" 
-            $label.Value = "Unknown"
-            break;  
-        }
-    }
+    param($reportPath)
     
+    Write-Host "##vso[task.addattachment type=Distributedtask.Core.Summary;name=SonarQube Analysis Report;]$reportPath"
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SummaryReport/ReportBuilder.ps1]: I guess we're only expecting one analysis report per build definition so attachment name doesn't need to be disambiguated e.g. by the project name.
I wonder what would happen if a user tried including multiple begin-build-end tasks sequence in the same build definition..

**Comment by duncanpMS** [comment]
LGTM

File: Tasks/SonarQubePostTest/SonarQubeMetrics.ps1 (hunk 11 above)
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: Calling WaitForAnalysisToComplete on an early SQ will throw a friendly error message. Currently this is only possible for the build breaker. In case of the summary report the old "legacy" report is uploaded if SQ version is 5.2 or lower so WaitForAnalysisToComplete won't even be called. 

But I understand your underlying question - as I call more and more APIs with different version requirements there should be a generic mechanism to handle the version check. I'll refactor fetching the metrics next time I need to call a metric requiring a version check.

File: Tasks/SonarQubePostTest/SonarQubeMetrics.ps1 (hunk 12 above)
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: Currently there are around 130 metrics defined. Note that users can define their own metrics as well, which is something that I haven't yet tested (not sure how popular this is).

File: Tasks/SonarQubePostTest/SonarQubeMetrics.ps1 (hunk 11 above)
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: Ok. According to the docs for _api/qualitygates_ it's been available since 5.4. What happens if the user is using 5.3?

Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SonarQubeMetrics.ps1]: Good catch! Actually calling api/qualitygates/project_status using the analysisId parameter is supported since 5.3 whereas projectId  / projectKey have been introduced in 5.4. I have used api/qualitygates/project_status with analysisId for the build breaker and this was tested against SonarQube 5.3;

File: Tasks/SonarQubePostTest/SummaryReport/ReportBuilder.ps1 (hunk 16 above)
Comment (bgavrilMS) [inline Tasks/SonarQubePostTest/SummaryReport/ReportBuilder.ps1]: I just tested and uploading the summary twice does not error out. Only one summary is displayed. I will make some changes to reset the cached values such as analysis ID in the pre build task to ensure 2 analysis can be run.

**Comment by HSAR** [comment]
LGTM
